### PR TITLE
[FrameworkBundle] Add debug:autoconfigure command

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | #26295  
| License       | MIT
| Doc PR        | todo

*SEE for a complete output and more insight* #28730

As far as I'm concerned, i've seen @ro0NL's comments on the PR, I think this is ready and working, we can tweak it, but letting the work from @aaa2000 being dead it not the right call IMHO. This is the the raw commit rebased with master and pushed.

### CODE REVIEW

**Comment by OskarStark** [comment]
Please use short array syntax 😊

**Comment by Simperfit** [comment]
cc @chalasr

**Comment by Simperfit** [comment]
AppVeyor failure seems not related

**Comment by Simperfit** [comment]
cc @ro0NL @chalasr what can we do about this ? Do we refator the whole thing or just let this work die ?

File: src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        $cloner = new VarCloner();
+        $cliDumper = new CliDumper(null, null, AbstractDumper::DUMP_LIGHT_ARRAY);
+
+        return $cliDumper->dump($cloner->cloneVar(current($tagAttribute)), true);
+    }
+
+    /**
+     * @return ContainerBuilder
+     */
+    private function getContainerBuilder()
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php]: is the version from `ContainerDebugCommand::getContainerBuilder()` blocking for this implementation? extending from `getContainerBuilder` gives us things for free

File: src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            $io->newLine();
+            $io->table(['Option', 'Value'], $tableRows);
+        }
+    }
+
+    private function dumpMethodCall(ChildDefinition $autoconfiguredInstanceofItem)
+    {
+        $tagContainerBuilder = new ContainerBuilder();
+        foreach ($tagContainerBuilder->getServiceIds() as $serviceId) {
+            $tagContainerBuilder->removeDefinition($serviceId);
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php]: isnt $tagContainerBuilder empty already :thinking:

File: src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+            $tagContainerBuilder->removeAlias($serviceId);
+        }
+        $tagContainerBuilder->addDefinitions([$autoconfiguredInstanceofItem]);
+
+        $dumper = new YamlDumper($tagContainerBuilder);
+        preg_match('/calls\:\n((?: +- .+\n)+)/', $dumper->dump(), $matches);
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php]: no big fan  of this approach, i'd opt for passing e.g. `getMethodCalls()` and render plain text as needed. Eventually could be inlined in `execute()`

File: src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+        preg_match('/parameters\:\n((?: + .+\n)+)/', $dumper->dump(), $matches);
+
+        return preg_replace('/^\s+/m', '', $matches[1]);
+    }
+
+    private function dumpTagAttribute(array $tagAttribute)
+    {
+        $cloner = new VarCloner();
+        $cliDumper = new CliDumper(null, null, AbstractDumper::DUMP_LIGHT_ARRAY);
```
Comment (ro0NL) [inline src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php]: #28898 is meant for this :)

**Review by ro0NL — COMMENTED** [review]
im still :+1: for `debug:autoconfiguration`

File: src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php (hunk 2 above)
Comment (maxhelias) [reply src/Symfony/Bundle/FrameworkBundle/Command/DebugAutoconfigurationCommand.php]: I think it's to make sure. And we could refactor it to be save empty and clone it on each call, no ?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 5]
```diff
@@ -376,7 +376,9 @@ public function load(array $configs, ContainerBuilder $container)
         $container->registerForAutoconfiguration(LocaleAwareInterface::class)
             ->addTag('kernel.locale_aware');
         $container->registerForAutoconfiguration(ResetInterface::class)
-            ->addTag('kernel.reset', ['method' => 'reset']);
+            ->addTag('kernel.reset', ['method' => 'reset'])
+            ->addTag('kernel.reset2', ['method' => 'reset2'])
```
Comment (maxhelias) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Would not it be a forgotten ? It seems strange to me

**Review by maxhelias — COMMENTED** [review]
I was wondering if this kind of command existed, now i have the answer.

I'm totally 👍 for this feature.

I have some question

**Comment by Simperfit** [comment]
will try to fix this tomorrow.

**Comment by maxhelias** [comment]
Or maybe we can do like for #32584, add a option like `--auto` ?

**Comment by nicolas-grekas** [comment]
I checked out the PR and here are two things that should be improved to me:
- the output should be way more compact
- the `getContainerBuilder()` method in the command always compiles the container. Since this is a slow process, we'd better read from the dumped XML container (then, the command could extend from `ContainerDebugCommand` as does `DebugAutowiringCommand` btw). This means we first need to dump autoconfigured types in the XML since they're not right now.

@Simperfit still available to work on this? Otherwise don't hesitate to call for help.

**Comment by Simperfit** [comment]
@nicolas-grekas Thanks for the review, ill look into it this week.

**Comment by nicolas-grekas** [comment]
Anyone willing to take over this one?

**Comment by atailouloute** [comment]
@nicolas-grekas I can take over this if no one did.

**Comment by nicolas-grekas** [comment]
Sure, please do!

**Comment by atailouloute** [comment]
Should I create a new PR starting from this branch ? I don't think I have write access to Simperfit's repo

**Comment by nicolas-grekas** [comment]
Please create another PR.

**Comment by chalasr** [comment]
Closing in favor of #35033.

**Pull request closed without merging**